File: python/MCTS/MCTS.py
import random
import numpy as np
from .TreeNode import TreeNode


class MCTS(object):
    def __init__(self, policy_value_fn, c_puct=5, n_playround=5000):
        self._policy_value = policy_value_fn
        self._c_puct = c_puct
        self._n_playround = n_playround
        self._root = TreeNode()

    def _softmax(self, x):
        probs = np.exp(x - np.max(x))
        probs /= np.sum(probs)
        return probs

    def _evaluate(self, winner, current_player):
        if winner == -1:
            return 0
        else:
            return 2 * (winner == current_player) - 1

    def _playout(self, board):
        cur = self._root
        while not cur.is_leaf():
            action, cur = cur.select(self._c_puct)
            board.step(action)
        if not board.over:
            action_probs, leaf_value = self._policy_value(board)
            cur.expend(action_probs)
        else:
            leaf_value = self._evaluate(board.winner, board.current_player)
        cur.update(-leaf_value)

    def get_action_probs(self, board, tau=1e-3):
        for _ in range(self._n_playround):
            board_copy = board.copy()
            self._playout(board_copy)
            #assert board_copy.moves != board.moves
            #assert board_copy.availables != board.availables
        act_visits = [(act, node.n_visits)
                      for act, node in self._root.children.items()]
        acts, visits = zip(*act_visits)
        act_probs = self._softmax(np.log(np.array(visits) + 1e-10) / tau)
        return acts, act_probs

    def update_with_action(self, action):
        if action in self._root.children:
            self._root = self._root.children[action]
            self._root.parent = None
        else:
            self._root = TreeNode()

    def depth(self):
        return self._depth(self._root)

    def _depth(self, node):
        if node.is_leaf():
            return 1
        max_depth = 0
        for action, node in node.children.items():
            child_depth = self._depth(node)
            if child_depth > max_depth:
                max_depth = child_depth
        return max_depth + 1

class _MCTS(object):
    def __init__(self, c_puct=5, n_playround=5000, limit=1000):
        self.max_round = limit
        self._c_puct = c_puct
        self._n_playround = n_playround
        self._root = TreeNode()

    def _evaluate(self, winner, current_player):
        if winner == -1:
            return 0
        else:
            return 2 * (winner == current_player) - 1

    def _playout(self, board):
        cur = self._root
        while not cur.is_leaf():
            action, cur = cur.select(self._c_puct)
            board.step(action)
        if not board.over:
            action_prob = np.ones(board.spaces) / board.spaces
            #assert len(board.availables) == board.spaces
            action_probs = zip(board.availables, action_prob)
            cur.expend(action_probs)
        leaf_value = self._evaluate_rollout(board)
        cur.update(-leaf_value)

    def _evaluate_rollout(self, board):
        player = board.current_player
        round = 0
        while not board.over and round < self.max_round:
            action = random.choice(board.availables)
            board.step(action)
            round += 1
        return self._evaluate(board.winner, player)

    def get_action(self, board):
        for _ in range(self._n_playround):
            board_copy = board.copy()
            self._playout(board_copy)
        return max(self._root.children.items(),
                   key=lambda act_node: act_node[1].n_visits)[0]

    def update_with_action(self, action):
        if action in self._root.children:
            self._root = self._root.children[action]
            self._root._parent = None
        else:
            self._root = TreeNode()

    def depth(self):
        return self._depth(self._root)

    def _depth(self, node):
        if node.is_leaf():
            return 1
        max_depth = 0
        for action, node in node.children.items():
            child_depth = self._depth(node)
            if child_depth > max_depth:
                max_depth = child_depth
        return max_depth + 1

File: src/MCTS/MCTS.cpp
//
// Created by Clytie on 2018/9/7.
//

#include "MCTS.h"

MCTS::MCTS(double c_puct, unsigned int n_playround, unsigned int limit)
        : c_pust(c_puct), n_playround(n_playround), max_round(limit) {
    root = new TreeNode(nullptr, 1.0);
}

void MCTS::playout(Board & board) {
    TreeNode * cur = root;
    while (!cur->is_leaf()) {
        pair<unsigned int, TreeNode *> best = cur->select(c_pust);
        cur = best.second;
        board.step(best.first);
    }
    if (!board.over()) {
        vector<double> probs(board.spaces, 1.0 / board.spaces);
        //assert(board.availables.size() == probs.size());
        vector<pair<unsigned int, double> > action_probs = utils::zip<unsigned int, double>(board.availables, probs);
        cur->expend(action_probs);
    }
    /* check
    if (!cur->is_root()) {
        cout << cur->parent->children.size() << " " << depth(cur) << " ";
        for (const auto & child : cur->parent->children) {
            if (child.second == cur) {
                cout << "first " << child.first << " ";
            }
        }
    }
     */
    double leaf_value = evaluate_rollout(board);

    cur->update(-leaf_value);
}

int MCTS::evaluate_rollout(Board & board) {
    unsigned int player = board.current_player();
    int round = 0;
    while (!board.over() && round < max_round) {
        auto i = 0;
        auto action_index = random() % board.spaces;
        for (typename list<unsigned int>::const_iterator itr = board.availables.begin(); itr != board.availables.end(); ++itr, ++i) {
            if (i == action_index) {
                board.step(*itr);
                break;
            }
        }
        ++round;
    }
    int value = 0;
    if (board.winner != -1) {
        value = 2 * (board.winner == player) - 1;
    }
    return value;
}

unsigned int MCTS::get_action(Board & board) { //每个状态(board)下都重建一棵蒙特卡罗树
    for (int i = 0; i < n_playround; ++i) {
        Board board_copy = board.copy();
        playout(board_copy);
    }
    vector<unsigned int> max_visits;
    unsigned int max_n_visits = 0;
    for (const auto & child : root->children) {
        if (child.second->n_visits > max_n_visits) {
            max_n_visits = child.second->n_visits;
            vector<unsigned int>().swap(max_visits);
            max_visits.emplace_back(child.first);
        } else if (child.second->n_visits == max_n_visits) {
            max_visits.emplace_back(child.first);
        }
    }
    return max_visits[random() % max_visits.size()];
}

void MCTS::update_with_action(unsigned int last_action) {
    if (last_action == (unsigned int)-1) {
        root = new TreeNode(nullptr, 1.0);
        return;
    }
    unordered_map<unsigned int, TreeNode *>::const_iterator itr = root->children.find(last_action);
    if (itr != root->children.end()) {
        root = itr->second;
        root->parent = nullptr;
    } else {
        root = new TreeNode(nullptr, 1.0);
    }
}

unsigned int MCTS::depth(TreeNode * node) {
    if (node->is_leaf()) {
        return 1;
    }
    unsigned int max_depth = 0;
    for (const auto & child : node->children) {
        unsigned int child_depth = depth(child.second);
        if (child_depth > max_depth) {
            max_depth = child_depth;
        }
    }
    return max_depth + 1;
}

unsigned int MCTS::depth() {
    return depth(root);
}
